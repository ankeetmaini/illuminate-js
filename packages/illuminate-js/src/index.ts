import {
    TokenObject,
    TokenTypes,
    tokenize,
    Definition,
    Plugin,
    addPlugin,
    highlight,
    addLanguage,
    getLanguage
} from './illuminate';

export {
    TokenObject,
    TokenTypes,
    tokenize,
    Definition,
    Plugin,
    addPlugin,
    highlight,
    addLanguage,
    getLanguage
};
