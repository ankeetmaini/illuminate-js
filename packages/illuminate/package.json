{
  "name": "illuminate-js",
  "version": "0.3.0",
  "description": "Syntax Highlighting for NodeJS",
  "main": "lib/index.js",
  "scripts": {
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vkbansal/illuminate-js.git"
  },
  "author": "Vivek Kumar Bansal <contact@vkbansal.me>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/vkbansal/illuminate-js/issues"
  },
  "homepage": "https://github.com/vkbansal/illuminate-js#readme"
}
