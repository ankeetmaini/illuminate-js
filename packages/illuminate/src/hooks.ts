interface Hooks {
    [key: string]: Set<(...args: any[]) => void>
}

const hooks: Hooks = {};

export function add(name: string, callback: (...args: any[]) => void) {
    if (typeof name !== 'string' || name.length < 1) {
        throw new Error('Name must be string of length > 1');
    }

    if (typeof callback !== 'function') {
        throw new Error(`hooks.add expects a function to be passed as a callback but ${typeof callback}:${callback} given`);
    }

    if (!hooks[name]) hooks[name] = new Set();

    hooks[name].add(callback);
}

export function run(name: string, env: object) {
    const callbacks = hooks[name];

    if (!callbacks || !callbacks.size) {
        return;
    }

    callbacks.forEach((callback) => callback(env));
}

export default { add, run };
