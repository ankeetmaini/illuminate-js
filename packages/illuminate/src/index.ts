import {
    TokenObject,
    TokenTypes,
    tokenize,
    Definition,
    Plugin,
    addPlugin,
    highlight
} from './illuminate';

export { TokenObject, TokenTypes, tokenize, Definition, Plugin, addPlugin, highlight };
