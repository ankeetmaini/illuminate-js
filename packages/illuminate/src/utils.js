/* eslint-disable import/prefer-default-export */
import Token from './Token';

export function encode(tokens) {
    if (tokens instanceof Token) {
        return new Token(tokens.type, encode(tokens.content), tokens.alias);
    } else if (Array.isArray(tokens)) {
        return tokens.map(encode);
    }

    return tokens
            .replace(/&/g, '&amp;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;')
            .replace(/\u00a0/g, ' ');
}
