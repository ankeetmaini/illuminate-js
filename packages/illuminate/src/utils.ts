import { Token } from './Token';

export type SingleOrArray<T> = T | Array<T>;

export function encode(tokens: SingleOrArray<Token | string>): SingleOrArray<Token | string> {
    if (tokens instanceof Token) {
        return new Token(tokens.type, <string>encode(tokens.content), tokens.alias);
    } else if (Array.isArray(tokens)) {
        return <Array<Token | string>>tokens.map(encode);
    }

    return tokens
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/\u00a0/g, ' ');
}
