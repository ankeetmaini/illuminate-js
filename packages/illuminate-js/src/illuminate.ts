import { Token } from './Token';
import { add as addHook, run as runHook, reset as resetHooks } from './hooks';
import { encode } from './utils';

export interface TokenObject {
    pattern: RegExp;
    inside?: Definition;
    lookbehind?: boolean;
    greedy?: boolean;
    alias?: string;
}

export type Tokens = RegExp | TokenObject | Array<RegExp | TokenObject>;

export type Definition = Map<string, Tokens | Map<string, Tokens>>;

export type Plugin = (a: typeof addHook) => void;

export const languages = new Map<string, Definition>();

/**
 * Converts the code to an `Array<string | Token>`, which can be used to write a render.
 * For example, `react-illuminate` uses this to tokenize the code string and render it using react components.
 *
 * @param text    The code to be tokenized
 * @param grammar The definition to be used for tokenizing
 *
 * @private
 *
 * @example
 * import { tokenize } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * const code = `
 * ...
 * `;
 *
 * tokenize(code, javascript); // returns Array<string | Token>
 */
export function tokenize(text: string, grammar: Definition): Array<string | Token> {
    if (!(grammar instanceof Map)) {
        throw new Error('A grammar must be a Map');
    }

    const startArray: Array<string | Token> = [text];

    const rest = grammar.get('rest');

    if (rest instanceof Map) {
        for (const [token, def] of rest as Definition) {
            grammar.set(token, def);
        }

        //TODO: check if we need to shallow copy the grammar
        grammar.delete('rest');
    }

    tokenloop: for (const [token, def] of grammar) {
        const patterns = Array.isArray(def) ? def : [def];

        for (let j = 0; j < patterns.length; ++j) {
            const pattern = patterns[j];
            let lookbehindLength: number = 0;

            let regex = pattern instanceof RegExp ? pattern : (pattern as TokenObject).pattern;

            const { inside, lookbehind, alias, greedy } = <TokenObject>pattern;

            if (greedy && !regex.global) {
                const flags = (regex.toString().match(/[imuy]*$/) || [''])[0];

                regex = RegExp(regex.source, `${flags}g`);
            }

            for (let i = 0; i < startArray.length; i++) {
                // Donâ€™t cache length as it changes during the loop
                const node = startArray[i];

                // Something went terribly wrong, ABORT, ABORT!
                if (startArray.length > text.length) {
                    break tokenloop;
                }

                if (node instanceof Token) {
                    continue;
                }

                regex.lastIndex = 0;

                let match: RegExpExecArray | string | null = regex.exec(node);

                if (!match) {
                    continue;
                }

                if (lookbehind) {
                    lookbehindLength = match[1].length;
                }

                const stringFrom: number = match.index - 1 + lookbehindLength;

                match = match[0].slice(lookbehindLength);

                const stringTo = stringFrom + match.length;
                const before = node.slice(0, stringFrom + 1);
                const after = node.slice(stringTo + 1);

                const inserts = [];

                if (before) {
                    inserts.push(before);
                }

                const wrapped = new Token(token, inside ? tokenize(match, inside) : match, alias);

                inserts.push(wrapped);

                if (after) {
                    inserts.push(after);
                }

                startArray.splice(i, 1, ...inserts);
            }
        }
    }

    return startArray;
}

export interface HighlightEnv {
    code: string;
    highlightedCode: string;
    grammar: Definition;
    language: string;
}

/**
 * Highlights the given code. Returns the highlighted code in `string` format.
 *
 * @param text      The code string that is to be highlighted.
 * @param language  Name of the definition to be used. This name corresponds to
 *                  the name given to the definition using `addLanguage`.
 *
 * @example
 * import { addLanguage, highlight } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * addLanguage('javascript', javascript);
 * addLanguage('js', javascript);
 *
 * const code = `
 * ...
 * `;
 *
 * highlight(code, 'js');
 */
export function highlight(text: string, language: string): string {
    const grammar = languages.get(language);

    if (!grammar) {
        throw new Error(
            `Your are trying to use ${language} language but its definition could not be found. Please make sure to use "addLanguage" to add your definition`
        );
    }

    let env: HighlightEnv = {
        grammar,
        language,
        code: text,
        highlightedCode: ''
    };

    env = <HighlightEnv>runHook('before-highlight', env) || env;

    const tokens = tokenize(env.code, env.grammar); // Specific for plugins
    const highlightedCode = Token.stringify(encode(tokens), language);

    env = Object.assign({}, env, { highlightedCode: highlightedCode }) || env;
    env = <HighlightEnv>runHook('after-highlight', env) || env;

    return env.highlightedCode; // Specific for plugins
}

/**
 * Helper to add a plugin.
 *
 * @param plugin The plugin to be added.
 *               See [Plugins](./#/plugins/) for more details.
 *
 * @example
 * import { addPlugin } from 'illuminate-js';
 * import { showLanguage } from 'illuminate-js/lib/plugins/showLanguage';
 *
 * addPlugin(showLanguage);
 */
export function addPlugin(plugin: Plugin) {
    if (typeof plugin !== 'function') {
        throw new Error('Given Plugin must be a function');
    }

    plugin(addHook);
}

/**
 * Removes all the plugins.
 *
 * @example
 * import { resetPlugins } from 'illuminate-js';
 *
 * resetPlugins();
 */
export function resetPlugins() {
    resetHooks();
}

/**
 * Add a language definition to a illuminate.
 * You can add same definition multiple times with different names
 *
 * @param name  The name of the language being added.
 *              This name will be used by `highlight` function.
 * @param def   The language `Definition` to be added.
 *
 * @example
 * import { addLanguage } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * addLanguage('javascript', javascript);
 * addLanguage('js', javascript);
 */
export function addLanguage(name: string, def: Definition) {
    languages.set(name, def);
}

/**
 * Returns the language definition which was added using `addLanguage`.
 * Returns the `Definition` if found, else returns `undefined`.
 *
 * @param name The name of the language definition
 *
 * @private
 *
 * @example
 * import { addLanguage, getLanguage } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * addLanguage('javascript', javascript);
 * addLanguage('js', javascript);
 *
 * getLanguage('js'); // returns javascript definition
 */
export function getLanguage(name: string): Definition | undefined {
    return languages.get(name);
}
