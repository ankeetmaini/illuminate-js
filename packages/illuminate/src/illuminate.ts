import { Token } from './Token';
import { add as addHook, run as runHook } from './hooks';
import { encode } from './utils';

export interface TokenObject {
    pattern: RegExp;
    inside?: Map<string, TokenTypes>;
    lookbehind?: boolean;
    greedy?: boolean;
    alias?: string;
}

export type TokenTypes = RegExp | TokenObject | Array<RegExp | TokenObject>;

export type Definition = Map<string, TokenTypes>;

export type Plugin = (a: typeof addHook) => void;

export function tokenize(text: string, grammar: Definition): Array<string | Token> {
    if (!(grammar instanceof Map)) {
        throw new Error('A grammar must be a Map');
    }

    const startArray: Array<string | Token> = [text];

    const rest = grammar.get('rest');

    if (rest instanceof Map) {
        for (const [token, def] of rest as Definition) {
            grammar.set(token, def);
        }

        //TODO: check if we need to shallow copy the grammar
        grammar.delete('rest');
    }

    tokenloop: for (const [token, def] of grammar) {
        const patterns = Array.isArray(def) ? def : [def];

        for (let j = 0; j < patterns.length; ++j) {
            const pattern = patterns[j];
            let lookbehindLength = 0;

            let regex = pattern instanceof RegExp ? pattern : pattern.pattern;

            const { inside, lookbehind, alias, greedy } = <TokenObject>pattern;

            if (greedy && !regex.global) {
                const flags = (regex.toString().match(/[imuy]*$/) || [''])[0];

                regex = RegExp(regex.source, `${flags}g`);
            }

            for (let i = 0; i < startArray.length; i++) {
                // Donâ€™t cache length as it changes during the loop
                const node = startArray[i];

                // Something went terribly wrong, ABORT, ABORT!
                if (startArray.length > text.length) {
                    break tokenloop;
                }

                if (node instanceof Token) {
                    continue;
                }

                regex.lastIndex = 0;

                let match: RegExpExecArray | string | null = regex.exec(node);

                if (!match) {
                    continue;
                }

                if (lookbehind) {
                    lookbehindLength = match[1].length;
                }

                const stringFrom = match.index - 1 + lookbehindLength;

                match = match[0].slice(lookbehindLength);

                const stringTo = stringFrom + match.length;
                const before = node.slice(0, stringFrom + 1);
                const after = node.slice(stringTo + 1);

                const inserts = [];

                if (before) {
                    inserts.push(before);
                }

                const wrapped = new Token(token, inside ? tokenize(match, inside) : match, alias);

                inserts.push(wrapped);

                if (after) {
                    inserts.push(after);
                }

                startArray.splice(i, 1, ...inserts);
            }
        }
    }

    return startArray;
}

export interface HighlightEnv {
    code: string;
    highlightedCode: string;
    grammar: Definition;
    language: string;
}

export function highlight(text: string, grammar: Definition, language: string): string {
    const env: HighlightEnv = {
        grammar,
        language,
        code: text,
        highlightedCode: ''
    };

    runHook('before-highlight', env);

    const tokens = tokenize(env.code, env.grammar); // Specific for plugins
    const highlightedCode = Token.stringify(encode(tokens), language);

    env.highlightedCode = highlightedCode;
    runHook('after-highlight', env);

    return env.highlightedCode; // Specific for plugins
}

export function addPlugin(plugin: Plugin) {
    if (typeof plugin !== 'function') {
        throw new Error('Given Plugin must be a function');
    }

    plugin(addHook);
}
