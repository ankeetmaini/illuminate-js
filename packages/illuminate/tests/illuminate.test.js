import { highlight, getLanguage } from '../src/illuminate';
import Definition from '../src/Definition';

describe('illuminate', () => {
    describe('getLanguage', () => {
        it('should return a language', () => {
            expect(getLanguage('clike')).toBeInstanceOf(Definition);
        });

        it('should return false when language not found', () => {
            expect(getLanguage('foo')).toBe(false);
        });
    });

    describe('highlight', () => {
        it('should highlight given text', () => {
            expect(highlight('if (a > b)', 'clike')).toMatchSnapshot();
        });
    });
});
