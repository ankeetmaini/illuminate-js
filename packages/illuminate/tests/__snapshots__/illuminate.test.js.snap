exports[`illuminate highlight should highlight given text 1`] = `"<span class=\"token keyword keyword-if\" >if</span> <span class=\"token punctuation brackets-parentheses\" >(</span>a <span class=\"token operator\" >&gt;</span> b<span class=\"token punctuation brackets-parentheses\" >)</span>"`;
